File: plugins/plugin-farcaster/typescript/actions/sendCast.ts
import {
  type Action,
  type ActionExample,
  createUniqueUuid,
  type IAgentRuntime,
  type Memory,
  ModelType,
  type State,
} from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import type { FarcasterService } from "../services/FarcasterService";
import { FARCASTER_SERVICE_NAME } from "../types";

const spec = requireActionSpec("SEND_CAST");

export const sendCastAction: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,
  examples: (spec.examples ?? []) as ActionExample[][],

  validate: async (runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const text = message.content.text?.toLowerCase() || "";
    const keywords = ["post", "cast", "share", "announce", "farcaster"];

    const hasKeyword = keywords.some((keyword) => text.includes(keyword));

    const service = runtime.getService(FARCASTER_SERVICE_NAME) as FarcasterService;
    const isServiceAvailable = !!service?.getCastService(runtime.agentId);

    return hasKeyword && isServiceAvailable;
  },

  handler: async (runtime: IAgentRuntime, message: Memory, state?: State) => {
    try {
      const service = runtime.getService(FARCASTER_SERVICE_NAME) as FarcasterService;
      const postService = service?.getCastService(runtime.agentId);

      if (!postService) {
        runtime.logger.error("[SEND_CAST] PostService not available");
        return { success: false, error: "PostService not available" };
      }

      let castContent = "";

      if (state?.castContent) {
        castContent = state.castContent as string;
      } else {
        const prompt = `Based on this request: "${message.content.text}", generate a concise Farcaster cast (max 320 characters). Be engaging and use appropriate hashtags if relevant.`;

        const response = await runtime.useModel(ModelType.TEXT_LARGE, { prompt });
        castContent = typeof response === "string" ? response : String(response);
      }

      if (castContent.length > 320) {
        castContent = `${castContent.substring(0, 317)}...`;
      }

      const cast = await postService.createCast({
        agentId: runtime.agentId,
        roomId: createUniqueUuid(runtime, "farcaster-timeline"),
        text: castContent,
      });

      runtime.logger.info(`[SEND_CAST] Successfully posted cast: ${cast.id}`);

      await runtime.createMemory(
        {
          agentId: runtime.agentId,
          roomId: cast.roomId,
          entityId: runtime.agentId,
          content: {
            text: castContent,
            source: "farcaster",
            metadata: {
              castHash: String(cast.metadata?.castHash || ""),
              action: "SEND_CAST",
            },
          },
          createdAt: cast.timestamp,
        },
        "messages"
      );
      return { success: true, text: `Posted cast: ${cast.id}` };
    } catch (error) {
      runtime.logger.error(
        "[SEND_CAST] Error posting cast:",
        typeof error === "string" ? error : (error as Error).message
      );
      return { success: false, error: (error as Error).message };
    }
  },
};

File: plugins/plugin-farcaster/python/elizaos_plugin_farcaster/actions/send_cast.py
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from elizaos_plugin_farcaster.service import FarcasterService
    from elizaos_plugin_farcaster.types import Cast


@dataclass
class ActionResult:
    success: bool
    text: str | None = None
    error: str | None = None
    data: dict[str, object] = field(default_factory=dict)


@dataclass
class ActionExample:
    name: str
    content: dict[str, object]


class SendCastAction:
    name = "SEND_CAST"
    similes = ["POST_CAST", "FARCASTER_POST", "CAST", "SHARE_ON_FARCASTER", "ANNOUNCE"]
    description = "Posts a cast (message) on Farcaster"

    examples: list[list[ActionExample]] = [
        [
            ActionExample(
                name="User",
                content={
                    "text": "Post 'Hello Farcaster!' to my timeline",
                    "source": "user",
                },
            ),
            ActionExample(
                name="Agent",
                content={
                    "text": "I've posted your message to Farcaster!",
                    "actions": ["SEND_CAST"],
                },
            ),
        ],
        [
            ActionExample(
                name="User",
                content={
                    "text": "Share this announcement on Farcaster",
                    "source": "user",
                },
            ),
            ActionExample(
                name="Agent",
                content={
                    "text": "Your announcement has been posted to Farcaster.",
                    "actions": ["SEND_CAST"],
                },
            ),
        ],
    ]

    @staticmethod
    def _has_cast_intent(text: str) -> bool:
        keywords = ["post", "cast", "share", "announce", "farcaster"]
        lower = text.lower()
        return any(keyword in lower for keyword in keywords)

    async def validate(
        self,
        message: dict[str, object],
        service: "FarcasterService | None" = None,
    ) -> bool:
        if not service or not service.is_running:
            return False

        content = message.get("content", {})
        text = str(content.get("text", "") if isinstance(content, dict) else "")

        return self._has_cast_intent(text)

    async def handler(
        self,
        message: dict[str, object],
        service: "FarcasterService",
        text_to_cast: str,
    ) -> ActionResult:
        if not service.is_running:
            return ActionResult(
                success=False,
                text="Farcaster service is not running",
                error="Service not started",
            )

        if len(text_to_cast) > 320:
            text_to_cast = text_to_cast[:317] + "..."

        casts: list[Cast] = await service.send_cast(text_to_cast, reply_to=None)

        if not casts:
            return ActionResult(
                success=False,
                text="Failed to send cast",
                error="No cast returned",
            )

        cast = casts[0]
        return ActionResult(
            success=True,
            text="Cast posted successfully!",
            data={
                "cast_hash": cast.hash,
                "text": cast.text,
                "author_fid": cast.author_fid,
            },
        )
